### ISSUE TITLE
`git-init` should specify to run as root when it needs it.

### ISSUE THREAD

**Issue by mattmoor** [tektoncd/pipeline#4711]
# Expected Behavior

A default nonroot version of `git-init` can pass e2e testing without issues.

# Actual Behavior

```
Error running git [init /workspace/sourcerepo]: exit status 1
/workspace/sourcerepo/.git: Permission denied
```

# Steps to Reproduce the Problem

I've been experimenting with an `apko` based replacement for https://github.com/tektoncd/pipeline/blob/main/images/git-init/Dockerfile

You can see it here: https://github.com/mattmoor/git-init/blob/main/.apko.yaml

Basically I replaced this base image with `ghcr.io/mattmoor/git-init`: https://github.com/tektoncd/pipeline/blob/30f4051a70e67373d042471c0cd5f546bc5a93b3/.ko.yaml#L4

At which point I see the output above when I `kubectl create` the following task:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: build-push-run-output-image-
spec:
  serviceAccountName: ""
  taskSpec:
    resources:
      inputs:
      - name: sourcerepo
        type: git
      outputs:
      - name: builtImage
        type: image
        targetPath: /workspace/sourcerepo
    steps:
    - name: build-and-push
      image: busybox
      script: |
        set -e
        cat <<EOF > $(inputs.resources.sourcerepo.path)/index.json
        {
        "schemaVersion": 2,
        "manifests": [
            {
            "mediaType": "application/vnd.oci.image.index.v1+json",
            "size": 314,
            "digest": "sha256:05f95b26ed10668b7183c1e2da98610e91372fa9f510046d4ce5812addad86b5"
            }
        ]
        }
    - name: echo
      image: busybox
      script: cat $(inputs.resources.sourcerepo.path)/index.json
  resources:
    inputs:
    - name: sourcerepo
      resourceSpec:
        type: git
        params:
        - name: revision
          value: v0.32.0
        - name: url
          value: https://github.com/GoogleContainerTools/skaffold
    outputs:
    - name: builtImage
      resourceSpec:
        type: image
        params:
        - name: url
          value: gcr.io/foo/bar
```

**Comment by sbwsg** [comment]
The patch at https://github.com/tektoncd/pipeline/pull/4717 seems to work with this custom `git-init` image.  It just adds an explicit `runAsUser` to the git pipelineresource container.

**Comment by mattmoor** [comment]
Looks like this didn't catch a few places in `TestYaml` and `TestExample`.

The two flavors of issue I see in those are:
1. `image: ko://.../git-init` and `script:` that needs root in `TestYamls`,
2. the SSH `creds-init` seems to be broken by this in `TestExamples`.

I'm chasing these in https://github.com/tektoncd/pipeline/pull/4758 now.

**Comment by mattmoor** [comment]
For the first issue, I am hoping a simple `runAsUser: 0` on the relevant steps does the trick (running now).

For the second issue, the root cause (I believe) is that the `ghcr.io/distroless/git` image doesn't have `/root`, so setting up `/root/.ssh` fails.  We should get this added to the distroless images because it is root's home directory in `/etc/passwd`, so it should be in the image.  Tekton could also choose to defensively `MkdirAll`.  I've started a thread about ensuring home dirs all exist in `apko` and I'll see how hard it would be to be somewhat defensive as well.

**Comment by abayer** [comment]
/reopen

I think we need a fix that doesn't require `runAsUser: 0` - we're going to break anyone using `git-init` directly as soon as they bump that to `v0.35.0`. Yeah, they can add `runAsUser: 0`, but I don't think making users add that explicitly is optimal.

**Comment by mattmoor** [comment]
If folks are invoking the `git-init` binary, then perhaps we could check the UID and warn for a release that they should `runAsUser: 0` if they need it, or they can preflight the change with `runAsUser: 65532`?

**Comment by dibyom** [comment]
/cc @pritidesai 

I think we need to update the v0.35 release notes letting users know that if they use git-init:latest, they need to run as root (https://github.com/tektoncd/pipeline/pull/4807/files)

**Comment by abayer** [comment]
So my immediate question is whether we think there's enough usage of `git-init:latest` out there in the wild to merit reverting the base image change ASAP. I've verified that all versions of the `git-clone` and `git-batch-merge` catalog tasks (which are the only tasks there using the `git-init` image) all have explicit default image versions specified, and #4717 means that git `PipelineResource`s work fine, so the only scenarios I can think of where someone could unknowingly pick up `git-init:v0.35.0` and get broken are either using one of those catalog tasks and overriding the `gitInitImage` parameter to use `git-init:latest`, or have written their own task using `git-init:latest`. 

I'm torn for my gut instinct on this - I hate breaking compatibility for existing tasks/pipelines/etc with a fiery passion, but that does feel like only a rather small portion of users are at all likely to hit this. 

Also - @dibyom, @lbernick - if I remember/understand correctly, we don't use the `git-init` Dockerfile any more, as of https://github.com/tektoncd/pipeline/commit/c84c0feaf096d0450d8594620bd9905424a50a96.

**Comment by dibyom** [comment]
Yeah I was initially thinking we'd have to revert and do a patch release but it seems like this will only break if you use git-init:latest. The existing git-clone task should work since it pins to v0.29 of git-init. So, I thinking just a breaking change note on the release notes should be sufficient.

**Comment by mattmoor** [comment]
I do see the following with a GCE PD-backed PVC 🤔 
```
$ kubectl logs debug -f
touch: /tmp/blah/asdf: Permission denied
total 16
drwx------    2 root     root         16384 Apr 28 20:01 lost+found
```

I'll dig a bit

**Comment by mattmoor** [comment]
I have a question out to one of the GKE security folks, but in the meantime...

Instead of having PVCs create this infectious `root` requirement, another option would be to wrap the `PVC` setup in a specific container that runs as `root` (explicitly) and broadens the permissions on `/workspace`.

If PVC-based `/workspace` requires `root`, then how is Tekton allowing ~any non-`root` builders in general? 🤔 

Given that this seems(?) PVC specific, it feels like something where the workaround is best if tightly isolated to that code IMO...

I'd be fine with rolling this back (or switching the `git` image to default-`root` for now), but part of trying to make `runAsUser: 0` explicit is being honest with ourselves about where we need `root`.  Everywhere seems like a **really** bad answer 😅

**Comment by vdemeester** [comment]
> ... on both EKS and GKE. I'll check PVCs now, but I would be very surprised if `root` requirements came from core K8s

Ah interesting. Ok then, it might be a problem for some k8s distribution but not all, I see.

@mattmoor I tend to agree with you, I just wonder if we should do the following before we switch to the full `nonroot` approach.

> Instead of having PVCs create this infectious `root` requirement, another option would be to wrap the `PVC` setup in a specific container that runs as `root` (explicitly) and broadens the permissions on `/workspace`.

> I'd be fine with rolling this back (or switching the `git` image to default-`root` for now), but part of trying to make `runAsUser: 0` explicit is being honest with ourselves about where we need `root`. Everywhere seems like a **really** bad answer

And I am all for that as well 👼🏼 I just want to make sure we don't get a flood of issues about that because we switch it in one release without making sure we wouldn't break half of our user base (I have no idea what it would represent).

**Comment by mattmoor** [comment]
Yeah, I’m definitely on board with avoiding breakages and scrambling.  I’m traveling this morning, but this afternoon, I can try to put together a version of distroless/git that uses root by default (likely tagged :root), so we still have the option to dump the Dockerfile and test the waters there, if we want to.

**Comment by mattmoor** [comment]
Ok, I have a couple PRs staged to add `ghcr.io/distroless/git:root` image, and once that lands, I can either PR here to use it, or we can simply roll things back.

If the main source of problems here is workspace / PVCs, does it make sense to split off a new issue to track overcoming the pervasive root requirement that seems to incur?

Are there other items we should break out into separate issues to try and make progress towards this?  Maybe having `git-init` itself emit a warning when run as `root`?  This could be noise if we can't distinguish between `:root` and `runAsUser: 0` 🤔

**Comment by mattmoor** [comment]
Ok we now have:
```yaml
# crane config ghcr.io/distroless/git:root | jq .
{
  "architecture": "amd64",
  "author": "github.com/chainguard-dev/apko",
  "created": "2022-04-29T19:02:04Z",
  "history": [
    {
      "author": "apko",
      "created": "2022-04-29T19:02:04Z",
      "created_by": "apko",
      "comment": "This is an apko single-layer image"
    }
  ],
  "os": "linux",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:d9d80e705806b426acb538e3a4b248cad7cc20fdc68ec81467422b4929cad935"
    ]
  },
  "config": {
    "Entrypoint": [
      "/usr/bin/git"
    ],
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
      "SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt"
    ]
  }
}
```

**Comment by abayer** [comment]
I just did a local run with `ghcr.io/distroless/git:root` as the base, then ran `examples/v1beta1/pipelineruns/pipelinerun.yaml` without `runAsUser: 0` in my kind setup, and it worked. (just to be safe, I did another run with `ghcr.io/distroless/git`, and it failed). So yeah, this seems Good Enough to me that we can just do a `v0.35.1` with this change and not roll back.

Also worth mentioning - in my kind setup, I don't get the permission denied error. Instead, I get the `unsafe repository ('/workspace/output' is owned by someone else)` error - which wouldn't show up if the `git config --add --global safe.directories (path)` call in `pkg/git/git.go` was _before_ the `git init` call. I'm opening another PR to flip the `git init` and `git config ...` calls, since I feel like that might create other issues in the future, and, well, I always should have put the `git config ...` call first. =)
